lighthouse-core/audits/third-party-summary.js

```diff
@@ -57,41 +76,90 @@ class ThirdPartySummary extends Audit {
-      const entity = thirdPartyWeb.getEntity(request.url);
-      if (!entity) continue;
-
-      const entityStats = entities.get(entity) || {...defaultEntityStat};
-      entityStats.transferSize += request.transferSize;
-      entities.set(entity, entityStats);
+      const urlStat = byURL.get(request.url) || {...defaultStat};
+      urlStat.transferSize += request.transferSize;
+      byURL.set(request.url, urlStat);
     }
 
     const jsURLs = BootupTime.getJavaScriptURLs(networkRecords);
 
     for (const task of mainThreadTasks) {
-      const attributeableURL = BootupTime.getAttributableURLForTask(task, jsURLs);
-      const entity = thirdPartyWeb.getEntity(attributeableURL);
-      if (!entity) continue;
+      const attributableURL = BootupTime.getAttributableURLForTask(task, jsURLs);
 
-      const entityStats = entities.get(entity) || {...defaultEntityStat};
+      const urlStat = byURL.get(attributableURL) || {...defaultStat};
       const taskDuration = task.selfTime * cpuMultiplier;
       // The amount of time spent on main thread is the sum of all durations.
-      entityStats.mainThreadTime += taskDuration;
+      urlStat.mainThreadTime += taskDuration;
       // The amount of time spent *blocking* on main thread is the sum of all time longer than 50ms.
       // Note that this is not totally equivalent to the TBT definition since it fails to account for FCP,
       // but a majority of third-party work occurs after FCP and should yield largely similar numbers.
-      entityStats.blockingTime += Math.max(taskDuration - 50, 0);
-      entities.set(entity, entityStats);
+      urlStat.blockingTime += Math.max(taskDuration - 50, 0);
+      byURL.set(attributableURL, urlStat);
+    }
+
+    // Map each URL's stat to a particular third party entity.
+    const /* Map<ThirdPartyEntity, string[]> */ urls = new Map();
+    for (const [url, urlStat] of byURL.entries()) {
+      const entity = thirdPartyWeb.getEntity(url);
+      if (!entity) {
+        byURL.delete(url);
+        continue;
+      }
+      const entityStat = byEntity.get(entity) || {...defaultStat};
+      entityStat.transferSize += urlStat.transferSize;
+      entityStat.mainThreadTime += urlStat.mainThreadTime;
+      entityStat.blockingTime += urlStat.blockingTime;
+      byEntity.set(entity, entityStat);
+
+      const entityURLs = urls.get(entity) || [];
+      entityURLs.push(url);
+      urls.set(entity, entityURLs);
     }
 
-    return entities;
+    return {byURL, byEntity, urls};
+  }
+
+  /**
+   * @param {ThirdPartyEntity} entity
+   * @param {{byEntity: Map<ThirdPartyEntity, Summary>, byURL: Map<string, Summary>, urls: Map<ThirdPartyEntity, string[]>}} summaries
+   * @param {Summary} stats
+   * @return {Array<!URLSummary>}
+   */
+  static getSubItems(entity, summaries, stats) {
+    const entityURLs = summaries.urls.get(entity) || [];
+    let items = entityURLs
+      .map(url => /** @type {URLSummary} */ ({url, ...summaries.byURL.get(url)}))
```

Comment:
type coercion caught my eye..

is it guaranteed `.byURL` will always return a value? if so, a comment here would be nice.

Reply:
I think so, but added a filter to be sure since there isn't a local guarantee